flag_engine/identities/models.py

```diff
@@ -1,28 +1,65 @@
 import datetime
 import typing
 import uuid
-from dataclasses import dataclass, field
 
+from pydantic import UUID4, BaseModel, Field, root_validator
+from pydantic_collections import BaseCollectionModel
+
+from flag_engine.features.models import FeatureStateModel
 from flag_engine.identities.traits.models import TraitModel
-from flag_engine.utils.collections import IdentityFeaturesList
 from flag_engine.utils.datetime import utcnow_with_tz
+from flag_engine.utils.exceptions import DuplicateFeatureState
+
+
+class IdentityFeaturesList(BaseCollectionModel[FeatureStateModel]):
+    @classmethod
+    def __get_validators__(
+        cls,
+    ) -> typing.Generator[typing.Callable[..., typing.Any], None, None]:
+        yield cls.validate
+        yield cls.ensure_unique_feature_ids
+
+    @classmethod
+    def ensure_unique_feature_ids(
+        cls, value: "IdentityFeaturesList"
+    ) -> "IdentityFeaturesList":
+        for i, feature_state in enumerate(value, start=1):
+            if feature_state.feature.id in [
+                feature_state.feature.id for feature_state in value[i:]
+            ]:
+                raise DuplicateFeatureState(
+                    f"Feature state for feature id={feature_state.feature.id} already exists"
+                )
+        return value
+
+    def append(self, feature_state: "FeatureStateModel") -> None:
+        self.ensure_unique_feature_ids([*self, feature_state])
+        super().append(feature_state)
 
 
-@dataclass
-class IdentityModel:
+class IdentityModel(BaseModel):
     identifier: str
     environment_api_key: str
-    created_date: datetime = field(default_factory=utcnow_with_tz)
-    identity_features: IdentityFeaturesList = field(
+    created_date: datetime.datetime = Field(default_factory=utcnow_with_tz)
+    identity_features: IdentityFeaturesList = Field(
         default_factory=IdentityFeaturesList
     )
-    identity_traits: typing.List[TraitModel] = field(default_factory=list)
-    identity_uuid: str = field(default_factory=uuid.uuid4)
-    django_id: int = None
+    identity_traits: typing.List[TraitModel] = Field(default_factory=list)
+    identity_uuid: UUID4 = Field(default_factory=uuid.uuid4)
+    django_id: typing.Optional[int] = None
+    composite_key: str = Field(default_factory=lambda: None)
 
-    @property
-    def composite_key(self) -> str:
-        return self.generate_composite_key(self.environment_api_key, self.identifier)
+    @root_validator(skip_on_failure=True)
```

Comment:
What's the benefit to using a concrete field vs a property? 

It feels kinda wrong that we're setting an attribute on the model in a validator, is this a valid pattern in pydantic or are we _slightly_ abusing the `root_validator` decorator here? Wouldn't this better live in an overridden `__init__` method (if we want to keep it as a concrete attr vs a property).